File: routers/roster.py
"""
Page roster router.
Manages assignment of Postiz integrations (pages) to projects,
and linking of Google Drive folders to pages.
"""

import os
import re

import httpx
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from services.roster import (
    get_page,
    list_all_pages,
    list_pages_for_project,
    load_roster,
    parse_drive_folder_id,
    remove_page,
    save_roster,
    set_page,
)

# Regex to strip emoji and other non-alphanumeric/space characters for dedup matching
_EMOJI_RE = re.compile(
    r"[\U0001F600-\U0001F64F\U0001F300-\U0001F5FF\U0001F680-\U0001F6FF"
    r"\U0001F1E0-\U0001F1FF\U00002702-\U000027B0\U000024C2-\U0001F251"
    r"\U0001F900-\U0001F9FF\U0001FA00-\U0001FA6F\U0001FA70-\U0001FAFF"
    r"\U00002600-\U000026FF\U0000FE0F\U0000200D]+",
    re.UNICODE,
)


def _normalize_name(name: str) -> str:
    """Normalize a page name for dedup: lowercase, strip emoji, collapse whitespace."""
    name = _EMOJI_RE.sub("", name)
    return re.sub(r"\s+", " ", name).lower().strip()

router = APIRouter()


# ── Roster CRUD ────────────────────────────────────────────────────────────


class UpdatePageRequest(BaseModel):
    project: str | None = None
    drive_folder_url: str | None = None


@router.get("/")
async def list_roster():
    """List all pages in the roster."""
    return {"pages": list_all_pages()}


@router.get("/project/{project_name}")
async def get_project_pages(project_name: str):
    """List pages assigned to a specific project, enriched with staging topic status."""
    from services.telegram import get_staging_group

    pages = list_pages_for_project(project_name)
    staging = get_staging_group()
    topics = staging.get("topics", {}) if staging else {}

    enriched = []
    for page in pages:
        iid = page.get("integration_id", "")
        topic_info = topics.get(iid)
        enriched.append({
            **page,
            "has_staging_topic": bool(topic_info and topic_info.get("topic_id")),
            "staging_topic_name": topic_info.get("topic_name") if topic_info else None,
        })

    return {"pages": enriched}


@router.put("/{integration_id}")
async def update_page(integration_id: str, req: UpdatePageRequest):
    """Assign or update a page in the roster."""
    data: dict = {}
    if req.project is not None:
        data["project"] = req.project
    if req.drive_folder_url is not None:
        data["drive_folder_url"] = req.drive_folder_url
        folder_id = parse_drive_folder_id(req.drive_folder_url)
        if req.drive_folder_url and not folder_id:
            raise HTTPException(
                status_code=400,
                detail="Could not parse Google Drive folder ID from URL",
            )
        data["drive_folder_id"] = folder_id
    entry = set_page(integration_id, data)
    return {"page": entry}


@router.delete("/{integration_id}")
async def delete_page(integration_id: str):
    """Remove a page from the roster."""
    removed = remove_page(integration_id)
    if not removed:
        raise HTTPException(status_code=404, detail="Page not found in roster")
    return {"deleted": True}


# ── Dedup ────────────────────────────────────────────────────────────────


@router.get("/duplicates")
async def find_duplicates():
    """Audit roster for duplicate page names.

    For each duplicate set, shows which integration_ids have:
    - staging topics (with topic_id)
    - inventory items (staged/pending/forwarded)
    So the user can decide which to keep.
    """
    from collections import defaultdict

    roster = load_roster()
    pages = roster["pages"]

    # Get staging topics and inventory data
    try:
        from services.telegram import get_staging_group, get_inventory
        staging = get_staging_group()
        staging_topics = staging.get("topics", {})
    except Exception:
        staging_topics = {}

    try:
        from services.telegram import load_config as load_tg_config
        tg_config = load_tg_config()
        all_inventory = tg_config.get("inventory", {})
    except Exception:
        all_inventory = {}

    # Group by normalized name (strips emoji, lowercases, collapses whitespace)
    by_name: dict[str, list[str]] = defaultdict(list)
    for ig_id, page in pages.items():
        key = _normalize_name(page.get("name") or ig_id)
        by_name[key].append(ig_id)

    duplicates = []
    for name, ids in sorted(by_name.items()):
        if len(ids) <= 1:
            continue

        entries = []
        for ig_id in ids:
            page = pages[ig_id]
            topic = staging_topics.get(ig_id, {})
            inv = all_inventory.get(ig_id, [])
            inv_pending = sum(1 for i in inv if not i.get("forwarded"))
            inv_forwarded = sum(1 for i in inv if i.get("forwarded"))

            entries.append({
                "integration_id": ig_id,
                "name": page.get("name", ""),
                "provider": page.get("provider", ""),
                "has_topic": bool(topic.get("topic_id")),
                "topic_id": topic.get("topic_id"),
                "topic_name": topic.get("topic_name", ""),
                "inventory_total": len(inv),
                "inventory_pending": inv_pending,
                "inventory_forwarded": inv_forwarded,
                "has_project": bool(page.get("project")),
                "has_drive": bool(page.get("drive_folder_id")),
            })

        duplicates.append({"name": name, "count": len(ids), "entries": entries})

    return {
        "total_pages": len(pages),
        "duplicate_names": len(duplicates),
        "duplicates": duplicates,
    }


@router.post("/dedup")
async def dedup_roster():
    """Remove duplicate roster entries that share the same name.

    For each duplicate set, keeps the entry with the most inventory/data.
    Merges inventory from removed entries into the kept one.
    Also cleans up duplicate staging topics.
    """
    from collections import defaultdict

    roster = load_roster()
    pages = roster["pages"]

    try:
        from services.telegram import (
            get_staging_group,
            remove_staging_topic,
            load_config as load_tg_config,
            save_config as save_tg_config,
        )
        staging = get_staging_group()
        staging_topics = staging.get("topics", {})
        tg_config = load_tg_config()
        all_inventory = tg_config.get("inventory", {})
    except Exception:
        staging_topics = {}
        tg_config = None
        all_inventory = {}

    by_name: dict[str, list[str]] = defaultdict(list)
    for ig_id, page in pages.items():
        key = _normalize_name(page.get("name") or ig_id)
        by_name[key].append(ig_id)

    removed_ids: list[str] = []
    removed_names: list[str] = []
    inventory_merged = 0

    for name, ids in by_name.items():
        if len(ids) <= 1:
            continue

        # Score: prefer entries with inventory > topic > project > drive
        def score(ig_id: str) -> int:
            s = 0
            inv = all_inventory.get(ig_id, [])
            s += len(inv) * 100  # inventory items matter most
            if ig_id in staging_topics:
                s += 50
            p = pages[ig_id]
            if p.get("project"):
                s += 10
            if p.get("drive_folder_id"):
                s += 5
            return s

        ids_sorted = sorted(ids, key=score, reverse=True)
        keep_id = ids_sorted[0]

        for dupe_id in ids_sorted[1:]:
            # Merge inventory from dupe into keeper
            dupe_inv = all_inventory.get(dupe_id, [])
            if dupe_inv and tg_config is not None:
                keep_inv = tg_config.get("inventory", {}).setdefault(keep_id, [])
                keep_inv.extend(dupe_inv)
                inventory_merged += len(dupe_inv)
                # Remove dupe inventory
                if dupe_id in tg_config.get("inventory", {}):
                    del tg_config["inventory"][dupe_id]

            removed_ids.append(dupe_id)
            removed_names.append(pages[dupe_id].get("name", dupe_id))
            remove_page(dupe_id)

    # Save merged inventory
    if tg_config is not None and inventory_merged > 0:
        save_tg_config(tg_config)

    # Clean up staging topics for removed IDs
    topic_removed = 0
    for rid in removed_ids:
        if rid in staging_topics:
            try:
                remove_staging_topic(rid)
                topic_removed += 1
            except Exception:
                pass

    return {
        "removed": len(removed_ids),
        "removed_names": removed_names,
        "inventory_merged": inventory_merged,
        "topics_cleaned": topic_removed,
        "remaining": len(list_all_pages()),
    }


# ── Sync with Postiz ──────────────────────────────────────────────────────


@router.get("/sync-notion/status")
async def notion_sync_status():
    """Check if Notion Master Pages sync is configured."""
    from services.notion_pages import is_configured
    return {"configured": is_configured()}


@router.post("/sync-notion")
async def sync_from_notion():
    """Pull account roster from Notion Master Pages DB.

    Notion is the source of truth for: username, signup email, password,
    forwarding address, poster name, group, account type, TikTok URL.
    Roster JSON keeps app-only fields (project, drive folder, CF email alias).
    """
    from services.notion_pages import is_configured, sync_into_roster

    if not is_configured():
        raise HTTPException(
            status_code=503,
            detail="Notion not configured — set NOTION_API_KEY and NOTION_PAGES_DB",
        )

    try:
        return await sync_into_roster()
    except httpx.HTTPStatusError as exc:
        raise HTTPException(
            status_code=exc.response.status_code,
            detail=f"Notion API error: {exc.response.text[:500]}",
        )
    except Exception as exc:
        raise HTTPException(status_code=500, detail=f"Notion sync failed: {exc}")


@router.post("/sync")
async def sync_integrations():
    """[DEPRECATED] Postiz sync — kept as alias to /sync-notion for backwards compat.

    Notion is now the source of truth. The legacy 'Sync from Postiz' button
    transparently routes here, which redirects to the Notion sync.
    """
    from services.notion_pages import is_configured, sync_into_roster

    if not is_configured():
        raise HTTPException(
            status_code=503,
            detail="Notion not configured — set NOTION_API_KEY and NOTION_PAGES_DB",
        )

    try:
        result = await sync_into_roster()
        # Match legacy response shape so old callers don't break
        return {
            "added": result.get("added", 0),
            "removed": 0,  # Notion-driven — we don't auto-remove anything
            "updated": result.get("updated", 0),
            "pages": result.get("pages", []),
        }
    except httpx.HTTPStatusError as exc:
        raise HTTPException(
            status_code=exc.response.status_code,
            detail=f"Notion API error: {exc.response.text[:500]}",
        )
    except Exception as exc:
        raise HTTPException(status_code=500, detail=f"Notion sync failed: {exc}")

File: rust/crates/api/src/routes/roster.rs
//! Page roster routes — parity port of `routers/roster.py`.
//!
//! CRUD over the page roster (project assignment + Drive folder mapping),
//! a duplicate-name audit + dedup, and the Notion Master Pages sync (Notion is
//! canonical for account identity/credentials; the roster keeps app-only
//! fields like project and drive_folder_id). `POST /api/roster/sync` is the
//! legacy Postiz-sync alias — it routes to the Notion sync with the legacy
//! response shape, exactly like the old app.

use std::collections::HashMap;

use axum::extract::{Path, State};
use axum::http::StatusCode;
use axum::routing::{get, post, put};
use axum::{Json, Router};
use clab_core::repo::roster as roster_repo;
use serde::Deserialize;
use serde_json::{json, Map, Value};

use super::sounds::HttpError;
use crate::state::AppState;

pub fn router() -> Router<AppState> {
    Router::new()
        // FastAPI served this at "/api/roster/" (trailing slash) and
        // redirected the bare path; axum treats them as distinct, so register
        // both against the same handler.
        .route("/api/roster/", get(list_roster))
        .route("/api/roster", get(list_roster))
        .route("/api/roster/project/{project_name}", get(get_project_pages))
        .route("/api/roster/duplicates", get(find_duplicates))
        .route("/api/roster/dedup", post(dedup_roster))
        .route("/api/roster/sync-notion", post(sync_from_notion))
        .route("/api/roster/sync", post(sync_legacy))
        .route("/api/roster/{integration_id}", put(update_page).delete(delete_page))
}

type ApiResult = Result<Json<Value>, HttpError>;

// ── CRUD ─────────────────────────────────────────────────────────────────────

/// GET /api/roster/ — every page in the roster.
async fn list_roster(State(st): State<AppState>) -> ApiResult {
    let pages = roster_repo::list_pages_full(&st.db).await?;
    let pages: Vec<Value> = pages.iter().map(|p| p.to_value()).collect();
    Ok(Json(json!({ "pages": pages })))
}

/// GET /api/roster/project/{project_name} — project pages enriched with
/// staging-topic status.
async fn get_project_pages(
    State(st): State<AppState>,
    Path(project_name): Path<String>,
) -> ApiResult {
    let pages = roster_repo::list_pages_for_project(&st.db, &project_name).await?;
    let topics = roster_repo::staging_topics(&st.db).await?;

    let enriched: Vec<Value> = pages
        .iter()
        .map(|page| {
            let mut v = page.to_value();
            let topic = topics.get(&page.id);
            if let Some(obj) = v.as_object_mut() {
                obj.insert("has_staging_topic".into(), json!(topic.is_some()));
                obj.insert(
                    "staging_topic_name".into(),
                    topic.map(|(_, name)| json!(name)).unwrap_or(Value::Null),
                );
            }
            v
        })
        .collect();

    Ok(Json(json!({ "pages": enriched })))
}

#[derive(Deserialize)]
struct UpdatePageRequest {
    project: Option<String>,
    drive_folder_url: Option<String>,
}

/// PUT /api/roster/{integration_id} — assign or update a page.
async fn update_page(
    State(st): State<AppState>,
    Path(integration_id): Path<String>,
    Json(req): Json<UpdatePageRequest>,
) -> ApiResult {
    let mut data = Map::new();
    if let Some(project) = req.project {
        data.insert("project".into(), json!(project));
    }
    if let Some(url) = req.drive_folder_url {
        let folder_id = roster_repo::parse_drive_folder_id(&url);
        if !url.is_empty() && folder_id.is_none() {
            return Err(HttpError::bad_request(
                "Could not parse Google Drive folder ID from URL",
            ));
        }
        data.insert("drive_folder_url".into(), json!(url));
        data.insert(
            "drive_folder_id".into(),
            folder_id.map(Value::String).unwrap_or(Value::Null),
        );
    }
    let entry = roster_repo::set_page(&st.db, &integration_id, &data).await?;
    Ok(Json(json!({ "page": entry.to_value() })))
}

/// DELETE /api/roster/{integration_id}
async fn delete_page(
    State(st): State<AppState>,
    Path(integration_id): Path<String>,
) -> ApiResult {
    if !roster_repo::remove_page(&st.db, &integration_id).await? {
        return Err(HttpError::not_found("Page not found in roster"));
    }
    Ok(Json(json!({ "deleted": true })))
}

// ── Dedup ────────────────────────────────────────────────────────────────────

/// The old emoji-stripping regex's character classes, as a predicate. The
/// legacy `0x24C2..=0x1F251` mega-range subsumed several of the smaller
/// classes (dingbats, misc symbols, flags, variation selector), so the union
/// below is the deduplicated equivalent.
fn is_emoji(c: char) -> bool {
    matches!(u32::from(c),
        0x200D                // zero-width joiner
        | 0x24C2..=0x1F251    // enclosed chars … enclosed ideographs (legacy mega-range)
        | 0x1F300..=0x1F5FF   // symbols & pictographs
        | 0x1F600..=0x1F64F   // emoticons
        | 0x1F680..=0x1F6FF   // transport & map
        | 0x1F900..=0x1F9FF   // supplemental symbols
        | 0x1FA00..=0x1FAFF   // extended symbols
    )
}

/// Normalize a page name for dedup: strip emoji, collapse whitespace, lowercase.
fn normalize_name(name: &str) -> String {
    let stripped: String = name.chars().filter(|c| !is_emoji(*c)).collect();
    stripped
        .split_whitespace()
        .collect::<Vec<_>>()
        .join(" ")
        .to_lowercase()
        .trim()
        .to_string()
}

/// Group page indices by normalized name, preserving roster order.
fn group_by_name(pages: &[roster_repo::RosterPage]) -> Vec<(String, Vec<usize>)> {
    let mut order: Vec<String> = Vec::new();
    let mut groups: HashMap<String, Vec<usize>> = HashMap::new();
    for (idx, page) in pages.iter().enumerate() {
        let key = if page.name.is_empty() {
            normalize_name(&page.id)
        } else {
            normalize_name(&page.name)
        };
        if !groups.contains_key(&key) {
            order.push(key.clone());
        }
        groups.entry(key).or_default().push(idx);
    }
    order
        .into_iter()
        .map(|key| {
            let ids = groups.remove(&key).unwrap_or_default();
            (key, ids)
        })
        .collect()
}

/// GET /api/roster/duplicates — audit duplicate page names with enough
/// context (topics, inventory, project, drive) to pick a keeper.
async fn find_duplicates(State(st): State<AppState>) -> ApiResult {
    let pages = roster_repo::list_pages_full(&st.db).await?;
    let staging_topics = roster_repo::staging_topics(&st.db).await?;
    let inventory = roster_repo::inventory_counts(&st.db).await?;

    let mut by_name = group_by_name(&pages);
    by_name.sort_by(|a, b| a.0.cmp(&b.0));

    let mut duplicates: Vec<Value> = Vec::new();
    for (name, indices) in &by_name {
        if indices.len() <= 1 {
            continue;
        }
        let entries: Vec<Value> = indices
            .iter()
            .map(|&idx| {
                let page = &pages[idx];
                let topic = staging_topics.get(&page.id);
                let (total, pending, forwarded) =
                    inventory.get(&page.id).copied().unwrap_or((0, 0, 0));
                json!({
                    "integration_id": page.id,
                    "name": page.name,
                    "provider": page.provider.clone().unwrap_or_default(),
                    "has_topic": topic.is_some(),
                    "topic_id": topic.map(|(id, _)| json!(id)).unwrap_or(Value::Null),
                    "topic_name": topic.map(|(_, n)| n.clone()).unwrap_or_default(),
                    "inventory_total": total,
                    "inventory_pending": pending,
                    "inventory_forwarded": forwarded,
                    "has_project": page.project.as_deref().is_some_and(|p| !p.is_empty()),
                    "has_drive": page.drive_folder_id.as_deref().is_some_and(|d| !d.is_empty()),
                })
            })
            .collect();
        duplicates.push(json!({ "name": name, "count": indices.len(), "entries": entries }));
    }

    Ok(Json(json!({
        "total_pages": pages.len(),
        "duplicate_names": duplicates.len(),
        "duplicates": duplicates,
    })))
}

/// POST /api/roster/dedup — for each duplicate set keep the entry with the
/// most data (inventory ≫ topic > project > drive), merge inventory into the
/// keeper, remove the rest (their staging topics go with them).
async fn dedup_roster(State(st): State<AppState>) -> ApiResult {
    let pages = roster_repo::list_pages_full(&st.db).await?;
    let staging_topics = roster_repo::staging_topics(&st.db).await?;
    let inventory = roster_repo::inventory_counts(&st.db).await?;

    let score = |page: &roster_repo::RosterPage| -> i64 {
        let mut s = 0i64;
        s += inventory.get(&page.id).map(|(t, _, _)| *t).unwrap_or(0) * 100;
        if staging_topics.contains_key(&page.id) {
            s += 50;
        }
        if page.project.as_deref().is_some_and(|p| !p.is_empty()) {
            s += 10;
        }
        if page.drive_folder_id.as_deref().is_some_and(|d| !d.is_empty()) {
            s += 5;
        }
        s
    };

    let mut removed_ids: Vec<String> = Vec::new();
    let mut removed_names: Vec<String> = Vec::new();
    let mut inventory_merged = 0u64;
    let mut topics_cleaned = 0u32;

    for (_, indices) in group_by_name(&pages) {
        if indices.len() <= 1 {
            continue;
        }
        // Stable sort keeps roster order among ties, like Python's sorted().
        let mut sorted = indices.clone();
        sorted.sort_by_key(|&idx| std::cmp::Reverse(score(&pages[idx])));
        let keep_id = pages[sorted[0]].id.clone();

        for &idx in &sorted[1..] {
            let dupe = &pages[idx];
            inventory_merged += roster_repo::move_inventory(&st.db, &dupe.id, &keep_id).await?;
            if staging_topics.contains_key(&dupe.id) {
                topics_cleaned += 1;
            }
            removed_ids.push(dupe.id.clone());
            removed_names.push(if dupe.name.is_empty() {
                dupe.id.clone()
            } else {
                dupe.name.clone()
            });
        }
    }

    for dupe_id in &removed_ids {
        roster_repo::remove_page(&st.db, dupe_id).await?;
    }

    let remaining = roster_repo::list_pages_full(&st.db).await?.len();
    Ok(Json(json!({
        "removed": removed_ids.len(),
        "removed_names": removed_names,
        "inventory_merged": inventory_merged,
        "topics_cleaned": topics_cleaned,
        "remaining": remaining,
    })))
}

// ── Notion Master Pages sync ─────────────────────────────────────────────────

/// Run the sync: Notion rows upsert into the roster. Notion is canonical for
/// account identity/credentials/poster; app-only fields (project, drive
/// folder id, CF email routing fields) are preserved from the existing entry.
async fn run_notion_sync(st: &AppState) -> Result<Value, HttpError> {
    if !integrations::notion::pages_configured() {
        return Err(HttpError::unavailable(
            "Notion not configured — set NOTION_API_KEY and NOTION_PAGES_DB",
        ));
    }

    let rows = integrations::notion::fetch_all_pages().await.map_err(|e| {
        match e.downcast_ref::<integrations::notion::NotionApiError>() {
            Some(api) => HttpError::new(
                StatusCode::from_u16(api.status).unwrap_or(StatusCode::BAD_GATEWAY),
                format!("Notion API error: {}", api.body),
            ),
            None => HttpError::new(
                StatusCode::INTERNAL_SERVER_ERROR,
                format!("Notion sync failed: {e}"),
            ),
        }
    })?;

    let mut added = 0u32;
    let mut updated = 0u32;
    let mut errors: Vec<String> = Vec::new();

    for row in &rows {
        let existing = match roster_repo::get_page_full(&st.db, &row.integration_id).await {
            Ok(v) => v,
            Err(e) => {
                errors.push(format!("{}: {e}", row.name));
                continue;
            }
        };
        let existing_v = existing.as_ref().map(|p| p.to_value()).unwrap_or(Value::Null);
        let app_only = |key: &str| existing_v.get(key).cloned().unwrap_or(Value::Null);

        // Merge: Notion fields overwrite, app-only fields preserved.
        let merged: Map<String, Value> = json!({
            "name": row.name,
            "provider": row.provider,
            "tiktok_url": row.tiktok_url,
            "signup_email": row.signup_email,
            "fwd_address": row.fwd_address,
            "password": row.password,
            "poster_name": row.poster_name,
            "group": row.group,
            "group_label": row.group_label,
            "account_type": row.account_type,
            "notes": row.notes,
            "notion_page_id": row.notion_page_id,
            "source": "notion",
            "status": row.status,
            "pipeline": row.pipeline,
            "page_type": row.page_type,
            "sounds_reference": row.sounds_reference,
            "go_live_date": row.go_live_date,
            // drive_folder_url: prefer Notion if set, else existing
            "drive_folder_url": if row.drive_folder_url.is_empty() {
                app_only("drive_folder_url")
            } else {
                json!(row.drive_folder_url)
            },
            // App-only fields preserved from the existing entry:
            "project": app_only("project"),
            "drive_folder_id": app_only("drive_folder_id"),
            "email_alias": app_only("email_alias"),
            "email_rule_id": app_only("email_rule_id"),
            "fwd_destination": app_only("fwd_destination"),
        })
        .as_object()
        .cloned()
        .unwrap_or_default();

        match roster_repo::set_page(&st.db, &row.integration_id, &merged).await {
            Ok(_) => {
                if existing.is_some() {
                    updated += 1;
                } else {
                    added += 1;
                }
            }
            Err(e) => errors.push(format!("{}: {e}", row.name)),
        }
    }

    let pages = roster_repo::list_pages_full(&st.db).await?;
    let pages: Vec<Value> = pages.iter().map(|p| p.to_value()).collect();

    Ok(json!({
        "added": added,
        "updated": updated,
        "total_in_notion": rows.len(),
        "errors": errors,
        "pages": pages,
    }))
}

/// POST /api/roster/sync-notion
async fn sync_from_notion(State(st): State<AppState>) -> ApiResult {
    Ok(Json(run_notion_sync(&st).await?))
}

/// POST /api/roster/sync — legacy Postiz-sync alias; runs the Notion sync and
/// answers in the legacy shape so old callers don't break.
async fn sync_legacy(State(st): State<AppState>) -> ApiResult {
    let result = run_notion_sync(&st).await?;
    Ok(Json(json!({
        "added": result["added"],
        "removed": 0, // Notion-driven — we don't auto-remove anything
        "updated": result["updated"],
        "pages": result["pages"],
    })))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn normalize_strips_emoji_and_collapses() {
        assert_eq!(normalize_name("Cool Page 🎵🔥"), "cool page");
        assert_eq!(normalize_name("  Cool   PAGE "), "cool page");
        assert_eq!(normalize_name("plain"), "plain");
        // variation selector + ZWJ sequences vanish too
        assert_eq!(normalize_name("star\u{fe0f}\u{200d}name"), "starname");
    }

    #[test]
    fn emoji_predicate_covers_legacy_ranges() {
        for c in ['😀', '🎵', '🚀', '✂', '🤖', '☀'] {
            assert!(is_emoji(c), "{c}");
        }
        for c in ['a', 'Z', '0', '-', 'é'] {
            assert!(!is_emoji(c), "{c}");
        }
    }
}
